### ISSUE TITLE
Cannot render mermaid to docx and pdf simultaneously

### ISSUE THREAD

**Issue by VMTdeJong** [quarto-dev/quarto-cli#3066]
### Bug description

# Quarto-mermaid-docx

I'm trying to use mermaid to produce a diagram in html, pdf and docx output. When I use Ctrl Shift B / Render All Formats, I get the following error:

`
ERROR: TypeError: error sending request for url (http://localhost:9222/json/list): error trying to connect: tcp connect error: No connection could be made because the target machine actively refused it. (os error 10061)
`

Regarding the output format:
- If I remove docx from the output in the YML file, the book renders correctly in pdf and html, when I use render all formats. 
- If I remove pdf from the output in the YML file, the book renders correctly in docx and html, when I use render all formats. 
- If I select rendering to MS Word only, the book renders correctly.
- If I select rendering to pdf only, the book renders correctly.

Regarding mermaid:
- If I remove the mermaid diagram, the book renders to all output formats without errors.
- If I wrap the mermaid code in `::: {.content-visible unless-format="docx"} :::` or similar, I still get the error when I try to render to all formats

Version running:
- Quarto 1.2 
- "Elsbeth Geranium" Daily (896a0002, 2022-08-02) for Windows
- R version 4.0.5 (2021-03-31) -- "Shake and Throw"
- Platform: x86_64-w64-mingw32/x64 (64-bit)

The obvious solution for me is not to use the Render All Formats option. But it took me way too long to find out, so I thought this may be helpful to know.

index.qmd:
````
# Preface

```{mermaid}
flowchart TB

a(1.) 
b[2.]

a --> b
```

bla bla bla
````

_quarto.yml:
````
project:
  type: book

book:
  title: "Mermaid docx pdf test"
  chapters:
    - index.qmd

format:
  html:
    theme: cosmo
  pdf:
    documentclass: scrreprt
  docx: default
    
editor: source
````

Here's a repository, which essentially contains the snippets here:
https://github.com/VMTdeJong/Quarto-mermaid-docx

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [X] Please document the RStudio IDE version you're running (if applicable), by providing the value displayed in the "About RStudio" main menu dialog?
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by VMTdeJong** [comment]
I also tried the code on another pc, where the issue persists:

quarto 1.2
RStudio 2022.07.2 Build 576
R version 4.1.1 (2021-08-10) -- "Kick Things"
Platform: x86_64-w64-mingw32/x64 (64-bit)

**Comment by cscheid** [comment]
Can you give us a more specific quarto version, from `quarto --version`?

**Comment by VMTdeJong** [comment]
Version 1.2.247

**Comment by VMTdeJong** [comment]
and version 1.0.37

**Comment by cscheid** [comment]
@cderv Can you reproduce this on your side? This error seems to be coming from puppeteer/cri (`http://localhost:9222/json/list` looks like an API request for CRI to provide its feature list).

**Comment by brianmcdonald** [comment]
Same issue while using mermaid on Quarto 1.2.262 on html exports when pdf and epub exports enabled. The error occurs when running quarto publish gh-pages from a windows machine and also on CI.

**Comment by cderv** [comment]
@cscheid I can reproduce on my side. Here is the stack
````
TypeError: error sending request for url (http://localhost:9222/json/list): error trying to connect: tcp connect error: Aucune connexion n’a pu être établie car l’ordinateur cible l’a expressément refusée. (os error 10061)
    at async mainFetch (deno:ext/fetch/26_fetch.js:288:14)
    at async fetch (deno:ext/fetch/26_fetch.js:505:9)
    at async externalRequest (file:///C:/Users/chris/Documents/DEV_R/quarto-cli/src/core/cri/deno-cri/external-request.js:21:20)
    at async Module.List (file:///C:/Users/chris/Documents/DEV_R/quarto-cli/src/core/cri/deno-cri/devtools.js:44:18)
    at async Chrome._fetchDebuggerURL (file:///C:/Users/chris/Documents/DEV_R/quarto-cli/src/core/cri/deno-cri/chrome.js:230:25)
    at async Chrome._start (file:///C:/Users/chris/Documents/DEV_R/quarto-cli/src/core/cri/deno-cri/chrome.js:176:19)
Debugger session ended
Process exited with code 1
````

it seems to come from our version of `cri.js` for Deno. I'll try to have a closer look. If you have some pointers, it can surely be useful

**Comment by cderv** [comment]
I believe you need to pass some options when initializing `cdp()` after opening Chrome, at least the port. 
https://github.com/quarto-dev/quarto-cli/blob/8f324fe72b788cabc91480dc81d6616ccc8eebef/src/core/cri/cri.ts#L113-L114
````diff
     open: async (url: string) => {
-      client = await cdp();
+      client = await cdp({ port: port });
````

Otherwise, `new Chrome` is called without options, and default port will be used, which is 9222. 
https://github.com/quarto-dev/quarto-cli/blob/8f324fe72b788cabc91480dc81d6616ccc8eebef/src/core/cri/deno-cri/chrome.js#L86

This creates the issue as it seems when rendering two formats, not the same Chrome process is used, and a new port is used. At least in my test, this part is called twice 
https://github.com/quarto-dev/quarto-cli/blob/8f324fe72b788cabc91480dc81d6616ccc8eebef/src/core/cri/cri.ts#L90

First with 9222 then 5939 

So maybe the issue is that not the same Chrome process is used for both rendering, but it would makes sense to close it after each format rendering also. 

Anyway, that is the issue a applying this patch fix it
```diff
diff --git a/src/core/cri/cri.ts b/src/core/cri/cri.ts
index 762ad76d8..975624084 100644
--- a/src/core/cri/cri.ts
+++ b/src/core/cri/cri.ts
@@ -111,7 +111,7 @@ export async function criClient(appPath?: string, port?: number) {
     rawClient: () => client,
 
     open: async (url: string) => {
-      client = await cdp();
+      client = await cdp({ port: port });
       const { Network, Page } = client;
       await Network.enable();
       await Page.enable();
```

Maybe we need to pass more options to `cdp()` though... 

Can I let you make a proper PR ? If not in a hurry I can do that later this week probably. 

Hope it helps fixing this.

**Comment by cscheid** [comment]
@cderv I should be able to take it from here. Thank you!

**Comment by cderv** [comment]
Great! If you need some testing, you'll assign back to me or ping me. Don't hesitate.

**Comment by cscheid** [comment]
@cderv @VMTdeJong I think we fixed this to the v1.3 prerelease builds. If you could check to see if the bug persists, I'd appreciate it (and feel free to reopen in that case!)

**Comment by cderv** [comment]
Thanks Carlos! 

That was indeed the fix I found. Glad only this was needed. Thank you for the PR.

**Comment by VMTdeJong** [comment]
I tried version 1.3.30 and 1.3.26, but neither seems to entirely solve the issue for me. Currently I am using 1.3.30. 

When I started writing this post, the browser sometimes opened http://localhost:7250/index.html, which produced an error message; ERR_CONNECTION_REFUSED . Still, when I opened the _book directory the html, docx and pdf files were there. It's just that the html file looked like a page from the early 90's:

![image](https://user-images.githubusercontent.com/30228864/204222340-28f84dc3-e92b-4b5d-a0e5-759f936d15a3.png)

At other times only a html file was produced.

-----

As I am finishing writing this post, I can no longer reproduce the ERR_CONNECTION_REFUSED  message in Chrome, even though I had this multiple times earlier. I have no idea why. 

Now it sometimes works fine (although it no longer automatically opens one of the output files). Sometimes the viewer in RStudio shows a mermaid error message, but not always. When it does, the html file looks fine in Chrome. See image:

![image](https://user-images.githubusercontent.com/30228864/204225562-cc3ae197-49b0-42b7-9e66-7684586cd101.png)

But at other times I only get site_libs, index.html, search.json, see updated repository https://github.com/VMTdeJong/Quarto-mermaid-docx/tree/main/_book

I don't know why I get different output at different times. Deleting the .quarto and _book directories does not appear to (consistently) solve the issue.
